File: 2022/20/python/main.py
# --- Constants ---
INPUT_FILE = "../input/input.txt";
KEY = 811589153;
MIX_COUNT = 10;
OFFSET_X = 1000
OFFSET_Y = 2000
OFFSET_Z = 3000


# --- Mix the order of numbers ---
def mix(ls):
    _len = len(ls) - 1
    for _id in range(len(ls)):
        i = next((i for i, item in enumerate(ls) if item[0] == _id), -1)
        if ls[i][1] == 0: continue
        item = ls.pop(i)
        j = (i + item[1]) % _len
        if j < 0: j = _len + j
        if j == 0: j = _len
        ls.insert(j, item)

# --- Find and sum the 3 coordinates ---
def sum_coord(ls):
    i = next((i for i, item in enumerate(ls) if item[1] == 0), -1)
    _len = len(ls)
    x = ls[((i + OFFSET_X) % _len)][1]
    y = ls[((i + OFFSET_Y) % _len)][1]
    z = ls[((i + OFFSET_Z) % _len)][1]
    return x + y + z

# --- Read and parse the input file ---
with open(INPUT_FILE) as file:
    text = file.read().strip()

# --- Puzzle 1 ---
ls = [ (i, int(x)) for i, x in enumerate(text.split("\n")) ]
mix(ls)
total = sum_coord(ls)
print(f"1. The sum of the coordinates: {total:,d}")

# --- Puzzle 1 ---
ls = [ (i, int(x) * KEY) for i, x in enumerate(text.split("\n")) ]
for _ in range(MIX_COUNT): mix(ls)
total = sum_coord(ls)
print(f"2. The sum of the coordinates: {total:,d}")

File: 2022/20/rust/src/main.rs
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants ---
const INPUT_FILE: &'static str = "../input/input.txt";
const KEY: i64 = 811589153;
const MIX_COUNT: usize = 10;
const OFFSET_X: usize = 1_000;
const OFFSET_Y: usize = 2_000;
const OFFSET_Z: usize = 3_000;

#[derive(PartialEq, Clone)]
struct Item(usize, i64);

// --- Mix the order of numbers ---
fn mix (list: &mut Vec<Item>) {
    let len = (list.len() - 1) as i64;
    for id in 0..list.len() {
        let i = list.iter().position(|x| x.0 == id).unwrap();
        if list[i].1 == 0 { continue; }
        let item = list.remove(i);
        let mut j = (i as i64 + item.1) % len;
        if j < 0 { j = len + j; }
        if j == 0 { j = len; }
        list.insert(j as usize, item);
    }
}

// --- Find and sum the 3 coordinates ---
fn sum(list: &Vec<Item>) -> i64 {
    let i = list.iter().position(|x| x.1 == 0).unwrap();
    let len = list.len();
    let x: i64 = list[(i + OFFSET_X) % len].1 as i64;
    let y: i64 = list[(i + OFFSET_Y) % len].1 as i64;
    let z: i64 = list[(i + OFFSET_Z) % len].1 as i64;
    x + y + z
}

// --- Main method ---
fn main() -> std::io::Result<()> {
    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    let vec: Vec<i64> = BufReader::new(file).lines()
        .map(|line| line.unwrap().parse::<i64>().unwrap())
        .collect();

    // --- Puzzle 1 ---
    let mut list: Vec<Item> = vec.iter().enumerate().map(|(i, x)| Item(i, *x)).collect();
    mix(&mut list);
    let total = sum(&list);
    println!("1. The sum of the coordiantes: {}", total);

    // --- Puzzle 2 ---
    let mut list:Vec<Item> = vec.into_iter().enumerate().map(|(i, x)| Item(i, x * KEY))
        .collect();
    for _ in 0..MIX_COUNT {
        mix(&mut list);
    }
    let total = sum(&list);
    println!("2. The sum of the coordinates: {}", total);

    Ok(())
}
